KAFKA-16437 - Upgrade to Jakarta and Jetty 12 (KIP-1032)

cshannon opened the pull request:
This commit implements the changes for [KIP-1032](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1032%3A+Upgrade+to+Jakarta+and+JavaEE+10+in+Kafka+4.0). This updates Kafka to Jakarta specs, JavaEE 10 and Jetty 12. The changes here primarily effect Kafka Connect and MM2.

**Todo/Notes:**

1. I bumped the connect modules to JDK 17 but I also had to bump a couple other things that had a dependency on conect. The tools project depends on connect so that had to be bumped, and streams depends on tools so that needed to be bumped. This means we may need to separate some things if we don't want to enforce JDK 17 on streams.
2. There is an issue with a test in DedicatedMirrorIntegrationTest that I had to change for now that involves escaping characters and not quite sure what to do about it yet. The cause is the Servlet 6 spec changing what is allowed in the path. See: https://github.com/jetty/jetty.project/issues/11890
3. I had to configure the idle timeout in Jetty requests to match our request timeout so tests didn't fail. This was needed to fix the `ConnectWorkerIntegrationTest#testPollTimeoutExpiry()` test

Testing is being done by just using the existing tests for Connect and MM2 which should be sufficient.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

cshannon commented:
@gharris1727 and @C0urante - If either (or both) of you could take a look at some point and see what you think. There's no rush but I wanted to get this PR going now that trunk has been bumped to 4.0.0-SNAPSHOT. I am going to be out of town the rest of the week but I'll be back online middle of next week.

cshannon commented:
The checks on Java 8 and Java 11 fail because this PR requires JDK 17+

cshannon committed:
KAFKA-16437 - Upgrade to Jakarta and Jetty 12.x

This commit implements KIP-1032 and upgrades to Jakarta, JavaEE 10 and
also Jetty 12. Jetty 12 requires JDK 17 so the connect and MM2 runtimes
have been bumped to JDK 17.

See KIP-1032

cshannon committed:
Fix SSLUtilsTest

cshannon committed:
Merge branch 'trunk' into KAFKA-16437

gharris1727 commented on build.gradle:
```diff
@@ -318,6 +318,18 @@ subprojects {
     addParametersForTests(name, options)
+
+    // Todo: Do we need to separate some components to not require JDK 17? For example,
+    // streams had to be bumped becuase of the dependency on tools.
```
Streams should still be built and tested with Java 11. I think this means we're blocked by https://issues.apache.org/jira/browse/KAFKA-17857

Also we are going to probably conflict with https://issues.apache.org/jira/browse/KAFKA-17811, and we can add connect and mirror to that list of modules to build with 17 as the minimum.

cshannon replied:
I can just keep this PR open until we are unblocked and periodically update it by merging trunk into the PR to make sure I resolve any conflicts so not a huge deal.

cshannon committed:
Merge branch 'trunk' into KAFKA-16437

cshannon committed:
Merge branch 'trunk' into KAFKA-16437

chia7712 commented:
@cshannon Could you please rebase code to include #17522?

cshannon committed:
Merge branch 'trunk' into KAFKA-16437

cshannon commented:
@chia7712 - I merged the latest trunk and fixed the conflicts so this PR is now up to date

cshannon committed:
Merge branch 'trunk' into KAFKA-16437

cshannon committed:
Bump jetty to 12.0.15

mumrah commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -1071,7 +1071,7 @@ private void awaitStopTask(ConnectorTaskId taskId, long timeout) {
 
             if (!task.awaitStop(timeout)) {
-                log.error("Graceful stop of task {} failed.", task.id());
+                log.error(" Graceful stop of task {} failed.", task.id());
```
nit: did you mean to add this space?

cshannon replied:
Nope, that was definitely inadvertent so I will fix that.

mumrah commented on gradle/spotbugs-exclude.xml:
```diff
@@ -40,6 +40,12 @@ For a detailed description of spotbugs bug categories, see https://spotbugs.read
     </Match>
 
+    <Match>
+        <!-- Disable warnings about doProvileged, This has been deprecated
+            as of JDK 17 for removal, see https://openjdk.org/jeps/411 -->
+        <Bug pattern="DP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED"/>
```
This comment confused me at first. Can you include the description of `DP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED` to make it clear to the reader?

> Classloaders should only be created inside doPrivileged block

mumrah commented on gradle/spotbugs-exclude.xml:
```diff
@@ -562,4 +568,10 @@ For a detailed description of spotbugs bug categories, see https://spotbugs.read
 
+    <Match>
+        <!-- DateFormat is not used in more than one thread -->
+        <Class name="org.apache.kafka.tools.ReplicaVerificationTool$ReplicaBuffer"/>
+        <Bug pattern="STCAL_INVOKE_ON_STATIC_DATE_FORMAT_INSTANCE"/>
```
Did any relevant code change in this PR for this spotbugs exclusion?

cshannon replied:
I don't remember why I needed to add this because I first worked on this back in July, so I went back to take a look at the file and it looks like I can remove this block entirely because @chia7712 recently added the [same](https://github.com/apache/kafka/blob/0f2a96879030e1ccc575ed10b5cb113efa8200bb/gradle/spotbugs-exclude.xml#L550-L553) exclusion in https://github.com/apache/kafka/pull/17426

mumrah reviewed (commented):
Thanks for the patch @cshannon. Overall, I think this change looks good. A few small things inline.

For big PRs like this, we rely heavily on our tests to ensure we didn't break anything. Can you run the test suite on this PR a few times just to see if anything shakes out? You can trigger a new build by pushing an empty commit (`git commit --allow-empty`)

cshannon committed:
Address PR comments

cshannon committed:
Merge branch 'trunk' into KAFKA-16437

cshannon commented:
@mumrah - Thanks for the review, I pushed an update and I believe I addressed all 3 of your comments (as well as merging in the latest from trunk) and that kicked off a new build. I can also run tests a couple more times with an empty commit after the tests finish.

cshannon committed:
empty commit to run tests

cshannon committed:
Merge branch 'trunk' into KAFKA-16437

mumrah reviewed (approved):
I re-ran the tests three times, looks good. I'm approving this now, but will wait a few days before merging to see if @gharris1727 or @C0urante (or @chia7712) have anything to add.

cshannon commented:
> I re-ran the tests three times, looks good. I'm approving this now, but will wait a few days before merging to see if @gharris1727 or @C0urante (or @chia7712) have anything to add.

@mumrah - Sounds good, thanks for the review. 

I forgot to mention that I looked more into what to do about having to [change](https://github.com/apache/kafka/pull/16754/files#diff-ab24c09f7ecde101419a599c0ee3625a8f180911149a252792490a5618c7dc7cR283) the `DedicatedMirrorIntegrationTest` to prevent a failure because of the Servlet 6 and Jakarta EE10 spec change that no longer allows ambiguous characters anymore in the URL path. 

If we wanted to restore the previous behavior to allow it, we could [set](https://javadoc.jetty.org/jetty-12/org/eclipse/jetty/server/HttpConfiguration.html#setUriCompliance(org.eclipse.jetty.http.UriCompliance)) the UriComplianceMode to be [LEGACY](https://javadoc.jetty.org/jetty-12/org/eclipse/jetty/http/UriCompliance.html#LEGACY). See this comment: https://github.com/jetty/jetty.project/issues/11890#issuecomment-2156449534

However, I would say it's better to not allow it as the spec now says it's a violation so I think the current change I made is fine, but I guess it depends if there's use cases to allow it.

chia7712 commented on gradle/dependencies.gradle:
```diff
@@ -162,15 +163,16 @@ libs += [
   jaxAnnotationApi: "javax.annotation:javax.annotation-api:$versions.jaxAnnotation",
   jaxbApi: "javax.xml.bind:jaxb-api:$versions.jaxb",
-  jaxrsApi: "javax.ws.rs:javax.ws.rs-api:$versions.jaxrs",
+  jakartaRsApi: "jakarta.ws.rs:jakarta.ws.rs-api:$versions.jakartaRs",
```
Could you please remove unnecessary item from license file after this change?

https://github.com/apache/kafka/blob/633cab91c1d5ca80763103fa38fb778dcf545709/LICENSE-binary#L305

Also, the following rule seems to be unnecessary.

https://github.com/apache/kafka/blob/trunk/build.gradle#L2900

chia7712 commented on gradle/dependencies.gradle:
```diff
@@ -162,15 +163,16 @@ libs += [
   jaxbApi: "javax.xml.bind:jaxb-api:$versions.jaxb",
-  jaxrsApi: "javax.ws.rs:javax.ws.rs-api:$versions.jaxrs",
+  jakartaRsApi: "jakarta.ws.rs:jakarta.ws.rs-api:$versions.jakartaRs",
+  jakartaServletApi: "jakarta.servlet:jakarta.servlet-api:$versions.jakartaServlet",
```
`jetty` 12.0.15 is using `jakarta.servlet` 6.0.0, so maybe we can remove this unused declaration?

cshannon replied:
Yeah it is being pulled in transitively by the ee10 Jetty jars so we can probably remove it because it's tied to that version.

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java:
```diff
@@ -113,6 +113,7 @@ public <T, U> T completeOrForwardRequest(FutureCallback<T> cb,
                 }
                 String forwardUrl = uriBuilder.build().toString();
                 log.debug("Forwarding request {} {} {}", forwardUrl, method, body);
+                // TODO, we may need to set the reqest timeout as Idle timeout on the HttpClient
```
typo: `reqest` -> `request`

chia7712 replied:
This comment is valid to me, as the default idle timeout for HttpClient is 30 seconds, and the default request timeout is 90 seconds.

cshannon replied:
If we want to configure the idle timeout I guess the question is the best way to do so. I think we'd either need to pass it as an argument to the request() methods or add it to the AbstractConfig for RestClient so it has access to it when constructing the HttpClient. I guess there's also the question if it should always match the request timeout or it could be configured as a separate config. Maybe this could be done as a follow on task after this is merged.

chia7712 reviewed (commented):
@cshannon thanks for this patch. a couple of comments are left. PTAL

cshannon committed:
address comments

gharris1727 commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:
```diff
@@ -280,7 +280,12 @@ public void testMultiNodeCluster() throws Exception {
             final String a = "A";
             // Use a convoluted cluster name to ensure URL encoding/decoding works
-            final String b = "B- ._~:/?#[]@!$&'()*+;=\"<>%{}|\\^`618";
+            // TODO: Jetty 12 throws a 400 amiguous error with the previous test string here
+            // It seems like at least some encoded characters are not allowed anymore based on
+            // https://github.com/jetty/jetty.project/issues/11890#issuecomment-2156442947
+            // See https://github.com/jetty/jetty.project/issues/11890
+            //final String b = "B- ._~:/?#[]@!$&'()*+;=\"<>%{}|\\^`618";
+            final String b = "B";
```
Could you keep most of these suspicious characters, and just remove the ones which are ambiguous?
From reading the linked issue, it might just be the `/` character, and the remainder are still valid.

It looks like this test was sensitive enough to find the underlying change in the library, so I'd like to keep it sensitive in case something like this happens again.

cshannon replied:
So there's actually a few characters that are not allowed, control characters, backslash, etc. I went ahead and removed a few problem characters so it works now. I also updated the comments for the section. You can see the exact rules in the `Rejecting Suspicious Sequences` section as part of https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#uri-path-canonicalization

gharris1727 commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/util/SSLUtilsTest.java:
```diff
@@ -33,6 +36,22 @@
+    private Password truststorePassword;
+
+    @BeforeEach
+    public void bofore() throws Exception {
```
nit typo: bofore

gharris1727 reviewed (approved):
Thanks @cshannon for this change, thanks so much for getting this completed and working through the merge conflicts with all of the rest of the activity on trunk!

My comments are non-blocking, i'm happy to merge this as-is, but i'll let you address any other comments and merge tomorrow before feature freeze.

cshannon commented:
@gharris1727 and @chia7712 - I think I addressed all the comments so this can be merged. The only outstanding [one](https://github.com/apache/kafka/pull/16754#discussion_r1879098262) was if we wanted to set the idle timeout on the HttpClient to match the same thing as the request timeout (or even make it configurable separately) but I am still not quite sure if we should and I think that might be better as a follow on discussion and don't think it's a blocker. I can work on a PR for that depending on what we wanted to do (if anything)

cshannon committed:
empty commit to re-run tests

chia7712 commented:
@cshannon There is a flaky test causing a timeout, and we are going to disable it temporarily. See [#18138](https://github.com/apache/kafka/pull/18138).

chia7712 commented:
This pull request causes a build issue. I have opened [KAFKA-18235](https://issues.apache.org/jira/browse/KAFKA-18235) to address and fix it.
